types/src/delta_set.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+//! Parallel data aggregation uses a `Delta` op. Every delta is is a state key
+//! (for accessing the storage) and an operation: a partial function with a
+//! postcondition.
+
+use crate::state_store::state_key::StateKey;
+
+/// Specifies different delta partial function specifications.
+#[derive(Copy, Clone, Hash, PartialOrd, Ord, PartialEq, Eq)]
+pub enum DeltaOp {
+    /// Addition of `value` which overflows on `limit`.
+    Addition { value: u128, limit: u128 },
+    /// Subtraction of `value` which cannot go below zero.
+    Subtraction { value: u128 },
+}
+
+impl DeltaOp {
+    /// Returns optional result of delta application to `base` (None if
+    /// postocndition not satisfied).
+    pub fn apply_to(&self, base: u128) -> Option<u128> {
+        match self {
+            DeltaOp::Addition { value, limit } => addition(base, *value, *limit),
+            DeltaOp::Subtraction { value } => subtraction(base, *value),
+        }
+    }
+}
+
+/// Implements application of `Addition` to `base`.
+fn addition(base: u128, value: u128, limit: u128) -> Option<u128> {
+    if value > (limit - base) {
+        None
+    } else {
+        Some(base + value)
+    }
+}
+
+/// Implements application of `Subtraction` to `base`.
+fn subtraction(base: u128, value: u128) -> Option<u128> {
+    if value > base {
+        None
+    } else {
+        Some(base - value)
+    }
+}
+
+impl std::fmt::Debug for DeltaOp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DeltaOp::Addition { value, limit } => {
+                write!(f, "+{} ensures result <= {}", value, limit)
+            }
+            DeltaOp::Subtraction { value } => {
+                write!(f, "-{} ensures 0 <= result", value)
+            }
+        }
+    }
+}
+
+/// Serializes value after delta application.
+pub fn serialize(value: &u128) -> Vec<u8> {
+    bcs::to_bytes(value).expect("unexpected serialization error")
+}
+
+/// Deserializes value for delta application.
+pub fn deserialize(value_bytes: &[u8]) -> u128 {
```

Comment:
why do we need serde here?

Reply:
We don't - my idea was to have a standard API the aggregator native functions or executor can use. Do you think `u128::to_be_bytes()` would be better?

Reply:
for native functions/executor why it needs to be translated to raw bytes?

Reply:
We need to check storage when we read, which means calling `get_state_value()` which returns bytes. 

Alternatively we can wrap `StateViewCache`/`VersionedView` (those that interact with deltas in sequential/parallel execution) to have a special map for deltas. But this is too much for the first prototype and is easy to change later.